File: src/solver.js
// Treasure Trash — exhaustive analysis. The rooms are tiny, so we do not estimate
// anything: every reachable state is enumerated, and liveness is computed exactly.
//
// This is the small-board version of what a Sokoban solver approximates with deadlock
// tables (Junghanns & Schaeffer, Rolling Stone): rather than pattern-matching a 5x4
// window, we label the whole reachable graph live-or-dead and read the answers off it.

import { DIR_ORDER, MOVE, explain, isWon, stateKey, cloneState, cell } from './rules.js';
import { formatLurd } from './format.js';

/** Thrown by `analyze` when `maxStates` is exceeded, so a caller can tell a room that is too
 *  big to enumerate from one that is merely unsolvable. */
export class TooManyStates extends Error {
  constructor(limit) { super(`state graph exceeds ${limit} states`); this.limit = limit; }
}

/**
 * Enumerate the level's whole state graph.
 *
 * `opts.maxStates` bounds the enumeration and THROWS `TooManyStates` when the bound is passed.
 * It never truncates: a partial graph would report a wrong par and a wrong trap count, and
 * every claim downstream of this function is only worth anything because the graph is whole.
 * Unbounded by default — a caller that has not thought about the bound gets the exact answer
 * or an out-of-memory crash, never a quiet lie.
 *
 * Returns:
 *   states       Map key -> { state, depth, edges:[{dir,kind,to}] }
 *   minMoves     length of the shortest win, or null if unsolvable
 *   shortestLurd one canonical shortest solution (ties broken by DIR_ORDER)
 *   shortestCount how many distinct shortest action-sequences exist (>1 = unintended solves)
 *   dead         Set of keys from which no win is reachable
 *   traps        [{ lurd, dir, kind, buriesExit }] live -> dead transitions
 *   silentTraps  subset of traps whose action is a plain MOVE
 */
export function analyze(start, opts = {}) {
  const maxStates = opts.maxStates ?? Infinity;
  const states = new Map();
  const rootKey = stateKey(start);
  states.set(rootKey, { state: start, depth: 0, edges: [], prev: null, prevAction: null });

  // --- forward BFS over the full reachable graph
  // How often the exit itself refuses an action is counted here rather than in a second sweep:
  // the BFS already visits every state and every direction, and already has the reason in hand.
  let exitRefusals = 0;
  let frontier = [rootKey];
  while (frontier.length) {
    const next = [];
    for (const key of frontier) {
      const node = states.get(key);
      for (const dir of DIR_ORDER) {
        const r = explain(node.state, dir);
        if (!r.ok) { if (r.reason === 'exit') exitRefusals++; continue; }
        const k = stateKey(r.next);
        if (!states.has(k)) {
          if (states.size >= maxStates) throw new TooManyStates(maxStates);
          states.set(k, { state: r.next, depth: node.depth + 1, edges: [], prev: key, prevAction: { dir, kind: r.kind } });
          next.push(k);
        }
        node.edges.push({ dir, kind: r.kind, to: k });
      }
    }
    frontier = next;
  }

  // --- who has won: asked once, in Map order, because two passes over the graph asking the
  // same question of every board is the whole board read twice.
  const wins = [];
  for (const [key, node] of states) if (isWon(node.state)) wins.push(key);

  // --- liveness: reverse-reachability from every winning state
  const reverse = new Map();
  for (const [key, node] of states) for (const e of node.edges) {
    const back = reverse.get(e.to);
    if (back) back.push(key); else reverse.set(e.to, [key]);
  }

  const live = new Set(wins);
  const queue = [...wins];
  while (queue.length) {
    const k = queue.pop();
    for (const p of reverse.get(k) ?? []) if (!live.has(p)) { live.add(p); queue.push(p); }
  }
  const dead = new Set();
  for (const k of states.keys()) if (!live.has(k)) dead.add(k);

  // --- traps: a legal action that takes a live state to a dead one
  const traps = [];
  for (const [key, node] of states) {
    if (!live.has(key)) continue;
    for (const e of node.edges) {
      if (!dead.has(e.to)) continue;
      traps.push({
        lurd: formatLurd([...pathTo(states, key), { dir: e.dir, kind: e.kind }]),
        dir: e.dir, kind: e.kind,
      });
    }
  }
  return {
    states, wins, dead, traps, exitRefusals,
    ...shortestFrom(states, rootKey, wins, wins),
    silentTraps: traps.filter(t => t.kind === MOVE),
    reachable: states.size,
  };
}

/**
 * The shortest win, read off an already-labelled graph: how far, one canonical line, and how
 * many distinct ones there are.
 *
 * `order` is what breaks the tie — the canonical line is the first shortest win the search
 * REACHED. A caller that has relabelled the graph from a new root has to pass its own discovery
 * order; handed the old one it would report a different one of the tied solves for a room
 * nothing had changed about, and `:solve` lines would churn on a re-run.
 */
function shortestFrom(states, rootKey, wins, order) {
  let minMoves = null;
  for (const k of wins) {
    const d = states.get(k).depth;
    if (minMoves === null || d < minMoves) minMoves = d;
  }
  if (minMoves === null) return { minMoves: null, shortestLurd: null, shortestCount: 0 };
  const best = new Set(wins.filter(k => states.get(k).depth === minMoves));
  return {
    minMoves,
    shortestLurd: formatLurd(pathTo(states, order.find(k => best.has(k)))),
    shortestCount: countShortestPaths(states, rootKey, best, minMoves),
  };
}

/**
 * The same room again, with the raccoon starting somewhere else.
 *
 * Where he starts picks the ROOT of the search. It does not touch the board, and the board is
 * what decides every edge — so the graph `analyze` built is already the graph for every other
 * start, and all that changes is the distances measured through it.
 *
 * Returns a fresh analysis over the same `states`, or NULL when it cannot promise that: if the
 * new start reaches fewer states than the graph holds, then `dead`, `traps` and `exitRefusals`
 * were counted over states this room no longer has, and the caller has to pay for a real
 * `analyze`. `states` is re-labelled in place — depths and back-pointers belong to whoever
 * rooted it last.
 */
export function reroot(a, start) {
  const rootKey = stateKey(start);
  if (!a.states.has(rootKey)) return null;

  for (const node of a.states.values()) { node.depth = -1; node.prev = null; node.prevAction = null; }
  a.states.get(rootKey).depth = 0;
  // Discovery order, kept because the canonical solve is a TIE-BREAK on it: `analyze` reports
  // the first shortest win its own search reached, and a re-root that reported a different one
  // of the tied solves would rewrite `:solve` lines for rooms nothing had changed about.
  const order = [rootKey];
  let frontier = [rootKey];
  while (frontier.length) {
    const next = [];
    for (const key of frontier) {
      const node = a.states.get(key);
      for (const e of node.edges) {
        const to = a.states.get(e.to);
        if (to.depth !== -1) continue;
        to.depth = node.depth + 1; to.prev = key; to.prevAction = { dir: e.dir, kind: e.kind };
        order.push(e.to); next.push(e.to);
      }
    }
    frontier = next;
  }
  if (order.length !== a.states.size) return null;

  // Which boards are won is a property of the boards, and re-rooting does not touch one, so the
  // set `analyze` found still stands. Asking `isWon` again would cost two sweeps of every cell
  // of every state — more than the relabelling this function exists to do instead.
  return { ...a, ...shortestFrom(a.states, rootKey, a.wins, order) };
}

/**
 * The set of boards from which the room can no longer be won — everything `analyze` computes
 * about liveness, and nothing else.
 *
 * A GENERATOR, driven a slice at a time, because this runs in the page: the rooms this pack
 * is heading toward reach tens of thousands of states, and computing that in one go would
 * stall the frame it started on. Yields `{ scanned }` every `budget` expansions and RETURNS
 * the dead set.
 *
 * Where `analyze` keeps a cloned board per state so it can report paths and pars, this keeps
 * only keys and adjacency — boards are dropped as soon as they are expanded. That is what
 * makes it affordable to run for every room the player opens.
 */
export function* deadScan(start, opts = {}) {
  const budget = opts.budget ?? 1200;
  const adj = new Map();                       // key -> [key], the only thing retained
  const wins = [];
  const seen = new Set([stateKey(start)]);
  let frontier = [[stateKey(start), start]];
  let work = 0;

  while (frontier.length) {
    const next = [];
    for (const [key, s] of frontier) {
      const outs = [];
      for (const dir of DIR_ORDER) {
        const r = explain(s, dir);
        if (!r.ok) continue;
        const k = stateKey(r.next);
        outs.push(k);
        if (!seen.has(k)) { seen.add(k); next.push([k, r.next]); }
      }
      adj.set(key, outs);
      if (isWon(s)) wins.push(key);
      if (++work % budget === 0) yield { scanned: work };
    }
    frontier = next;                           // the previous layer's boards go out of scope
  }

  const rev = new Map();
  for (const [k, outs] of adj) for (const o of outs) {
    if (!rev.has(o)) rev.set(o, []);
    rev.get(o).push(k);
  }
  const live = new Set(wins);
  const stack = [...wins];
  while (stack.length) {
    const k = stack.pop();
    for (const p of rev.get(k) ?? []) if (!live.has(p)) { live.add(p); stack.push(p); }
    if (++work % budget === 0) yield { scanned: work };
  }

  const dead = new Set();
  for (const k of adj.keys()) if (!live.has(k)) dead.add(k);
  return dead;
}

function pathTo(states, key) {
  const out = [];
  for (let k = key; states.get(k).prev !== null; k = states.get(k).prev) out.push(states.get(k).prevAction);
  return out.reverse();
}

function countShortestPaths(states, rootKey, targets, minMoves) {
  const byDepth = [];
  for (const [k, n] of states) (byDepth[n.depth] ??= []).push(k);
  const ways = new Map([[rootKey, 1]]);
  for (let d = 0; d < minMoves; d++) {
    for (const k of byDepth[d] ?? []) {
      const w = ways.get(k) ?? 0;
      if (!w) continue;
      for (const e of states.get(k).edges) {
        if (states.get(e.to).depth !== d + 1) continue;   // stay on the shortest-path DAG
        ways.set(e.to, (ways.get(e.to) ?? 0) + w);
      }
    }
  }
  return [...targets].reduce((a, k) => a + (ways.get(k) ?? 0), 0);
}

/** Throws on the first disagreement between the declared solution and the engine. */
export function replay(start, actions) {
  let s = cloneState(start);
  const trace = [s];
  actions.forEach((a, i) => {
    const r = explain(s, a.dir);
    if (!r.ok) throw new Error(`move ${i + 1} (${a.kind} ${a.dir}): illegal — blocked by ${r.reason}`);
    if (r.kind !== a.kind) throw new Error(`move ${i + 1} (${a.dir}): declared ${a.kind}, board gives ${r.kind}`);
    s = r.next; trace.push(s);
  });
  return { final: s, trace };
}

File: engine/src/solver.rs
//! Exhaustive analysis, in Rust. A port of `src/solver.js`, which stays the engine of record.
//!
//! The rooms are tiny, so nothing here estimates: every reachable state is enumerated and
//! liveness is computed exactly. Same contract as the JS — the bound THROWS rather than
//! truncating, because a partial graph reports a wrong par and a wrong trap count and every
//! claim downstream is only worth anything because the graph is whole.
//!
//! What the protocol's `answer` carries is five numbers, and none of them depends on the order
//! the search reached things in. The canonical `:solve` DOES — it is a tie-break on discovery
//! order — and it is deliberately not in the contract, so it is not computed here. The states
//! are still held in a `Vec` in insertion order, indexed by a side table, which is what will
//! make adding it cost nothing later.

use crate::board::*;
use crate::rules::{explain, Outcome};
use std::collections::HashMap;

const DIR_ORDER: [u8; 4] = [b'u', b'd', b'l', b'r'];

pub struct Answer {
    pub par: Option<i32>,
    pub solves: u64,
    pub traps: usize,
    pub reachable: usize,
    pub exit_refusals: usize,
}

struct Node {
    state: State,
    depth: i32,
    edges: Vec<u32>,
}

/// Canonical state key, byte for byte the string `stateKey` builds — every lane of it is here
/// because dropping one is a SILENT bug, and the reasons are written out in `src/rules.js`.
///
/// The labelling of the multi-cell pieces is the lane most easily got wrong: the ids do not
/// determine the partition, so they are relabelled BY FIRST APPEARANCE in raster order. Keep the
/// raw ids instead and two boards that differ only in which couch is called which count as two
/// states, which shows up as a `reachable` that is too high and a par that is right by luck.
fn state_key(s: &State) -> Vec<u8> {
    let mut kinds = Vec::with_capacity((s.rows * (s.cols + 1)) as usize);
    let (mut pids, mut carts) = (Vec::new(), Vec::new());
    let (mut pid_seen, mut cart_seen): (Vec<u16>, Vec<u16>) = (Vec::new(), Vec::new());
    let label = |seen: &mut Vec<u16>, id: u16| -> u8 {
        match seen.iter().position(|&k| k == id) {
            Some(i) => i as u8,
            None => {
                seen.push(id);
                (seen.len() - 1) as u8
            }
        }
    };
    for y in 0..s.rows {
        if y > 0 {
            kinds.push(b'/');
        }
        for x in 0..s.cols {
            let c = s.at(x, y);
            let terrain: u16 = if c.water { 1 } else if c.bridge { 2 } else { 0 };
            let cart = if c.is_cart() { 1 } else { 0 };
            kinds.push((65 + (c.o as u16 * 3 + terrain) * 2 + cart) as u8);
            if c.pid != NO_ID {
                pids.push(65 + label(&mut pid_seen, c.pid));
            }
            if c.cart != NO_ID {
                carts.push(65 + label(&mut cart_seen, c.cart));
            }
        }
    }
    let mut key = kinds;
    key.push(b'|');
    key.extend(pids);
    key.push(b'|');
    key.extend(carts);
    key.push(b'|');
    key.extend(format!("{},{}", s.rac.0, s.rac.1).into_bytes());
    key
}

fn bags_left(s: &State) -> u32 {
    s.cells
        .iter()
        .map(|c| match c.o {
            BAG | CAN_FULL | WHEELIE | BIN => 1,
            STACK => 2,
            _ => 0,
        })
        .sum()
}

fn trash_held(s: &State) -> u32 {
    s.cells.iter().filter(|c| c.is_cart() && c.o == TRASH).count() as u32
}

fn is_won(s: &State) -> bool {
    bags_left(s) == 0 && trash_held(s) == 0 && s.at(s.rac.0, s.rac.1).exit
}

pub fn analyze(start: &State, max_states: usize) -> Result<Answer, String> {
    let mut index: HashMap<Vec<u8>, u32> = HashMap::new();
    let mut nodes: Vec<Node> = Vec::new();
    index.insert(state_key(start), 0);
    nodes.push(Node { state: start.clone(), depth: 0, edges: Vec::new() });

    // How often the exit itself refuses an action is counted here rather than in a second
    // sweep: the BFS already visits every state and every direction, and has the reason in hand.
    let mut exit_refusals = 0usize;
    let mut frontier = vec![0u32];
    while !frontier.is_empty() {
        let mut next = Vec::new();
        for &at in &frontier {
            let depth = nodes[at as usize].depth;
            // Collected first so the borrow of `nodes` ends before the graph grows. `explain`
            // hands back an owned board, so nothing here needs a copy of the one it came from.
            let mut got = Vec::with_capacity(4);
            for dir in DIR_ORDER {
                got.push(explain(&nodes[at as usize].state, dir)?);
            }
            for r in got {
                let ns = match r {
                    Outcome::No { reason } => {
                        if reason == "exit" {
                            exit_refusals += 1;
                        }
                        continue;
                    }
                    Outcome::Ok { next, .. } => next,
                };
                // Keyed once. Hashing a board is a scan of every cell of it, so asking twice
                // to look up and then insert is a second full read of the same board.
                let k = state_key(&ns);
                let to = if let Some(&i) = index.get(&k) {
                    i
                } else {
                    if nodes.len() >= max_states {
                        return Err(format!("state graph exceeds {max_states} states"));
                    }
                    let i = nodes.len() as u32;
                    index.insert(k, i);
                    nodes.push(Node { state: ns, depth: depth + 1, edges: Vec::new() });
                    next.push(i);
                    i
                };
                nodes[at as usize].edges.push(to);
            }
        }
        frontier = next;
    }

    let wins: Vec<u32> = (0..nodes.len() as u32)
        .filter(|&i| is_won(&nodes[i as usize].state))
        .collect();

    // Liveness: reverse-reachability from every winning state.
    let mut reverse: Vec<Vec<u32>> = vec![Vec::new(); nodes.len()];
    for (i, n) in nodes.iter().enumerate() {
        for &to in &n.edges {
            reverse[to as usize].push(i as u32);
        }
    }
    let mut live = vec![false; nodes.len()];
    let mut queue = wins.clone();
    for &w in &wins {
        live[w as usize] = true;
    }
    while let Some(k) = queue.pop() {
        for &p in &reverse[k as usize] {
            if !live[p as usize] {
                live[p as usize] = true;
                queue.push(p);
            }
        }
    }

    // A trap is a legal action that takes a live state to a dead one. Counted per EDGE, so two
    // directions from one board into the same dead board are two ways to lose, not one.
    let mut traps = 0usize;
    for (i, n) in nodes.iter().enumerate() {
        if !live[i] {
            continue;
        }
        traps += n.edges.iter().filter(|&&to| !live[to as usize]).count();
    }

    let par = wins.iter().map(|&w| nodes[w as usize].depth).min();
    let solves = match par {
        None => 0,
        Some(m) => count_shortest(&nodes, &wins, m),
    };

    Ok(Answer { par, solves, traps, reachable: nodes.len(), exit_refusals })
}

/// How many distinct shortest action-sequences win. Counted over the shortest-path DAG rather
/// than enumerated, so a room with thousands of tied lines costs the same as one with a single
/// line. More than one means the room has solves nobody intended.
fn count_shortest(nodes: &[Node], wins: &[u32], par: i32) -> u64 {
    let mut by_depth: Vec<Vec<u32>> = Vec::new();
    for (i, n) in nodes.iter().enumerate() {
        let d = n.depth as usize;
        if by_depth.len() <= d {
            by_depth.resize(d + 1, Vec::new());
        }
        by_depth[d].push(i as u32);
    }
    let mut ways = vec![0u64; nodes.len()];
    ways[0] = 1;
    for d in 0..par as usize {
        let Some(layer) = by_depth.get(d) else { break };
        for &k in layer {
            let w = ways[k as usize];
            if w == 0 {
                continue;
            }
            for &to in &nodes[k as usize].edges {
                if nodes[to as usize].depth == d as i32 + 1 {
                    ways[to as usize] += w;
                }
            }
        }
    }
    wins.iter()
        .filter(|&&w| nodes[w as usize].depth == par)
        .map(|&w| ways[w as usize])
        .sum()
}
